src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaTokenOptionsConfiguration.cs

```diff
@@ -0,0 +1,39 @@
+using System.Security.Cryptography;
+using Microsoft.Extensions.Options;
+using OrchardCore.Entities;
+using OrchardCore.Settings;
+
+namespace OrchardCore.Media.Processing
+{
+    public class MediaTokenOptionsConfiguration : IConfigureOptions<MediaTokenOptions>
+    {
+        private readonly ISiteService _siteService;
+
+        public MediaTokenOptionsConfiguration(ISiteService siteService)
+        {
+            _siteService = siteService;
+        }
+
+        public void Configure(MediaTokenOptions options)
+        {
+            var mediaTokenSettings = _siteService.GetSiteSettingsAsync()
+                .GetAwaiter().GetResult()
+                .As<MediaTokenSettings>();
+
+            if (mediaTokenSettings.HashKey == null)
+            {
+                var siteSettings = _siteService.LoadSiteSettingsAsync().GetAwaiter().GetResult();
+
+                var rng = RandomNumberGenerator.Create();
+
+                mediaTokenSettings.HashKey = new byte[64];
+                rng.GetBytes(mediaTokenSettings.HashKey);
+                siteSettings.Put(mediaTokenSettings);
+
+                _siteService.UpdateSiteSettingsAsync(siteSettings).GetAwaiter().GetResult();
+            }
+
+            options.HashKey = mediaTokenSettings.HashKey;
```

Comment:
@jtkech all felt wrong still.

So I made an options class, handled the migration generation there, and made the `IMediaTokenService` scoped.

Kept the feature event handler for refreshing the key

Reply:
@deanmarcussen 

Okay no problem i will work on it this night

As a reminder: Not a problem to use `GetAwaiter().GetResult()` for site settings, as i remember they are aleady get through a filter and then cached for the whole scope lifetime.

But as said above, really need to know the key needs to be in sync across instances serving the same we app